### PR TITLE
Move stable hasher out of rustc_middle

### PR DESCRIPTION
**Pull request by cjgillot**
Continuing the effort of dismantling `rustc_middle`,
`rustc_middle::ich` and half of `rustc_middle::middle` are moved into a dependency crate.

The new crate name is open for bikeshedding.

### CODE REVIEW

File: compiler/rustc_crate/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+use rustc_span::symbol::Symbol;
+
+pub mod codegen_fn_attrs;
+pub mod cstore;
+pub mod dependency_format;
+pub mod ich;
+pub mod limits;
+pub mod privacy;
+pub mod stability;
```
Comment (Mark-Simulacrum) [inline compiler/rustc_crate/src/lib.rs:21]: This seems like a very disjoint set of modules -- are they all interdependent? If not, it seems odd to move them into a new crate as a single group.

**Comment by Mark-Simulacrum** [comment]
We'll want to perf run this once the merge conflict is resolved.

File: compiler/rustc_crate/src/lib.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_crate/src/lib.rs:21]: I started by aiming at cstore, which is needed for ICH, and with what I could in rustc_middle::middle.
I welcome suggestions on the correct place for each module.

**Comment by cjgillot** [comment]
perf seems neutral.

**Comment by petrochenkov** [comment]
@cjgillot 
Could you explain what is the purpose of introducing `rustc_crate` if no dependencies on `rustc_middle` were removed and `rustc_middle` still depends on `rustc_crate`?
This doesn't introduce new parallelism and can only make sense if compilation time grows superlinearly with crate size.

**Comment by cjgillot** [comment]
I agree. Marking as draft until I can achieve dependency reduction.

**Pull request closed without merging**